planner/core/rule_column_pruning.go

```diff
@@ -151,10 +151,30 @@ func (ls *LogicalSort) PruneColumns(parentUsedCols []*expression.Column) error {
 	return child.PruneColumns(parentUsedCols)
 }
 
+// PruneColumns implements LogicalPlan interface.
+// If any expression can view as a constant in execution stage, such as correlated column, constant,
+// we do prune them. Note that we can't prune the expressions contain non-deterministic functions, such as rand().
+func (lt *LogicalTopN) PruneColumns(parentUsedCols []*expression.Column) error {
+	child := lt.children[0]
+	for i := len(lt.ByItems) - 1; i >= 0; i-- {
+		cols := expression.ExtractColumns(lt.ByItems[i].Expr)
+		if len(cols) == 0 {
+			if expression.IsMutableEffectsExpr(lt.ByItems[i].Expr) {
+				continue
+			}
+			lt.ByItems = append(lt.ByItems[:i], lt.ByItems[i+1:]...)
+		} else if lt.ByItems[i].Expr.GetType().Tp == mysql.TypeNull {
+			lt.ByItems = append(lt.ByItems[:i], lt.ByItems[i+1:]...)
+		} else {
+			parentUsedCols = append(parentUsedCols, cols...)
```

Comment:
Should we check whether the columns are in `parentUsedCols`?

Reply:
Both can. The redundant columns will not affect correctness.